itou/metabase/tables/utils.py

```diff
@@ -91,12 +91,9 @@ def get_choice(choices, key):
 
 
 def get_first_membership_join_date(memberships):
-    memberships = list(memberships.all())
-    # We have to do all this in python to benefit from prefetch_related.
+    # Prefetched memberships are sorted by -joined_at
```

Comment:
Le `order_by()` pourrais être modifié ou supprimé que personne irai voir dans cette fonction, donc pas super à l'aise d'avoir cette supposition.
Par contre le tri est effectivement coûteux pour rien, ce qu'on cherche au final c'est `min(joined_at)` donc soit le faire en python ici, ou alors peut-être essayer en forçant une sous-requête avec un `.annotate(date_inscription=SubQuery(...))`.
Le coup de la sous requête pourrais d'ailleurs peut-être aussi être essayer pour le `active_memberships` plutôt que _prefetch_ :shrug:, y a moyen que ça soit pas pire niveaux perf et ça nous éviterais du code :shrug:.

Reply:
Alors de manière assez incompréhensible, quand j'ai essayé de faire un annotate à la place, c'était plus long (pour les entreprises) ou infiniment trop long (pour les organisations).

Je vais ré-essayer d'écrire les annotations pour essayer de comprendre ce qui pose problème.

Reply:
Ce que tu me dis me choque pas pour un `.annotate(date_inscription=Min("members__joined_at"))` car il va y avoir un JOIN de la mort avec les utilisateurs, mais un `.annotate(date_inscription=SubQuery(CompanyMembership.objects.active().filter(company=OuterRef("pk").annotate(min=Min("joined_at").values("min"))))` (pas sûr des parenthèses mais tu vois l'idée :grin:) devrait je pense plutôt bien s'en sortir et ne pas être dépendant des autres annotations ou de la requête de base.

Reply:
```
first_membership_join_date=Subquery(
    CompanyMembership.objects.filter(company_id=OuterRef("id"))
    .values("company_id")
    .annotate(min=Min("joined_at"))
    .values("min")
),
active_membership_count=Subquery(
     CompanyMembership.objects.filter(company_id=OuterRef("id"))
     .active()
     .values("company_id")
     .annotate(count=Count("pk"))
     .values("count")
),
last_login=Subquery(
    CompanyMembership.objects.filter(company_id=OuterRef("id"))
    .values("company_id")
    .annotate(last_login=Max("user__last_login"))
    .values("last_login")
),
```
Ca fonctionne, mais c'est beaucoup plus long (similaire à l'annotation directe)

Je laisse comme c'est ici du coup :)